File: Leetcode/arrays-and-strings/sliding-window/maximum-average-subarray/maximumAverageSubarray.ts
function maximumAverageSubarray(nums: number[], k: number): number {
    let sum = 0;
    let max: number;

    for (let i = 0; i < k; i++ ) {
        sum += nums[i]
    }

    max = sum / k;

    for (let i = k; i < nums.length; i++ ) {
        sum += nums[i] - nums[i - k]
        max = Math.max(max, sum / k)
    }

    return max;
}

// Time: O(n);
// Space: O(n);

const result = maximumAverageSubarray([1, 12, -5, -6, 50, 3], 4);
console.log(result);

File: Leetcode/arrays-and-strings/sliding-window/maximum-average-subarray/maximumAverageSubarray.java
public double findMaxAverage(int[] nums, int k) {
    double current = 0;

    for (int i = 0; i < k; i++) {
        current += nums[i];
    }

    double answer = current;

    for (int i = k; i < nums.length; i++) {
        current += nums[i] - nums[i - k];
        answer = Math.max(answer, current);
    }

    return answer / k;
}
